File: src/compono/validator.py
"""Overflow validator: fonttools-based glyph advance-width measurement.

COMPONO_PLAN.md section 7. The core measurement/wrap/overflow functions are
pure and independently unit-testable with zero rendering involved — they
operate on a `FontMetrics` table, not a font file. `load_font_metrics` is the
one function that touches a real font file (via fontTools), reading actual
advance widths directly rather than rasterizing anything.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from fontTools.ttLib import TTFont  # type: ignore[import-untyped]  # no stubs published

FONTS_DIR = Path(__file__).parent / "fonts"

# Bundled/verified-present fonts this library will render text with. Anything
# outside this list is substituted for the template default or flagged.
# Empty until real font files are bundled (see fonts/.gitkeep) — resolve_safe_font
# returns None for every name until then, which is the correct/honest behavior.
SAFE_FONTS: dict[str, str] = {}


@dataclass(frozen=True)
class FontMetrics:
    """Per-character advance widths, in font units, normalized by `units_per_em`."""

    advance_widths: dict[str, int]
    units_per_em: int
    default_advance: int  # fallback for characters missing from the font's cmap


def load_font_metrics(font_path: Path) -> FontMetrics:
    """Read real glyph advance widths from a font file. No rendering required."""
    font = TTFont(font_path, lazy=True)
    try:
        cmap = font.getBestCmap()
        hmtx = font["hmtx"]
        units_per_em = font["head"].unitsPerEm

        advance_widths: dict[str, int] = {}
        for codepoint, glyph_name in cmap.items():
            advance_widths[chr(codepoint)] = hmtx[glyph_name][0]

        space_advance = advance_widths.get(" ")
        default_advance = space_advance if space_advance is not None else round(units_per_em * 0.5)

        return FontMetrics(
            advance_widths=advance_widths,
            units_per_em=units_per_em,
            default_advance=default_advance,
        )
    finally:
        font.close()


def resolve_safe_font(name: str) -> Path | None:
    """Look up a bundled/verified-present font by name. None if not in the allowlist."""
    filename = SAFE_FONTS.get(name)
    return FONTS_DIR / filename if filename else None


def _char_width_pt(char: str, metrics: FontMetrics, font_size_pt: float) -> float:
    raw = metrics.advance_widths.get(char, metrics.default_advance)
    return raw / metrics.units_per_em * font_size_pt


def measure_text_width_pt(text: str, metrics: FontMetrics, font_size_pt: float) -> float:
    """Sum measured advance widths for `text` at `font_size_pt`, in points."""
    return sum(_char_width_pt(c, metrics, font_size_pt) for c in text)


def wrap_lines(
    text: str,
    metrics: FontMetrics,
    font_size_pt: float,
    max_width_pt: float,
) -> list[str]:
    """Greedy line-wrap using measured advance widths (COMPONO_PLAN.md section 7)."""
    words = text.split()
    if not words:
        return []

    space_w = _char_width_pt(" ", metrics, font_size_pt)
    lines: list[str] = []
    current_words: list[str] = []
    current_width = 0.0

    for word in words:
        word_w = measure_text_width_pt(word, metrics, font_size_pt)
        candidate_width = word_w if not current_words else current_width + space_w + word_w
        if current_words and candidate_width > max_width_pt:
            lines.append(" ".join(current_words))
            current_words = [word]
            current_width = word_w
        else:
            current_words.append(word)
            current_width = candidate_width

    if current_words:
        lines.append(" ".join(current_words))
    return lines


@dataclass(frozen=True)
class OverflowReport:
    lines: list[str]
    line_count: int
    total_height_pt: float
    box_height_pt: float
    overflow: bool


def check_overflow(
    text: str,
    metrics: FontMetrics,
    font_size_pt: float,
    box_width_pt: float,
    box_height_pt: float,
    line_height_factor: float = 1.2,
) -> OverflowReport:
    """Greedy line-wrap -> line count * line-height -> compare to box height."""
    lines = wrap_lines(text, metrics, font_size_pt, box_width_pt)
    line_height_pt = font_size_pt * line_height_factor
    total_height_pt = len(lines) * line_height_pt
    return OverflowReport(
        lines=lines,
        line_count=len(lines),
        total_height_pt=total_height_pt,
        box_height_pt=box_height_pt,
        overflow=total_height_pt > box_height_pt,
    )


def build_overflow_error(
    slide: int,
    primitive_path: str,
    field: str,
    report: OverflowReport,
    font_size_pt: float,
) -> dict[str, object]:
    """Structured, actionable error — a fix, not just a diagnosis (COMPONO_PLAN.md section 8, item 3)."""
    excess_pt = report.total_height_pt - report.box_height_pt
    return {
        "slide": slide,
        "primitive": primitive_path,
        "field": field,
        "error": "overflow",
        "detail": (
            f"Text is ~{excess_pt:.0f}pt too tall for the box at font size {font_size_pt:g}pt "
            f"({report.line_count} lines)."
        ),
        "fix": "Shorten the text, reduce bullet/line count, or split into two slides.",
    }

File: packages/compono-js/src/validator.ts
/**
 * Overflow detection via real font metrics — TypeScript port of
 * src/compono/validator.py, backed by `fontkit` in place of `fontTools`.
 *
 * fontkit reads glyph advance widths directly from a font file (no
 * rendering), the same approach as the Python side's hmtx/cmap access.
 */

import { existsSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { openSync as openFontSync } from "fontkit";
import type { Template } from "./resolver.js";

export interface FontMetrics {
  advanceWidths: Map<string, number>;
  unitsPerEm: number;
  defaultAdvance: number;
}

const __dirname = dirname(fileURLToPath(import.meta.url));
export const FONTS_DIR = join(__dirname, "..", "fonts");

// Maps each stock template's `fontFamily` name (see templates/*.yaml) to a
// bundled font file used ONLY to measure overflow (real glyph advance
// widths via fontkit) — it is never written into the output .pptx/.docx.
// What actually renders in the deck is controlled solely by
// `template.fontFamily`, which PowerPoint/Word resolve against whatever's
// installed on the viewer's machine, same as any other OOXML font-name
// reference.
//
// Every stock template currently maps to the same bundled file (Open Sans,
// SIL OFL 1.1 — see fonts/OFL.txt) as an approximation: it's the one real
// font shipped with the package, so overflow math has genuine glyph widths
// to work with instead of skipping validation entirely.
export const SAFE_FONTS: Record<string, string> = {
  Calibri: "OpenSans-Regular.ttf",
  Georgia: "OpenSans-Regular.ttf",
  "Times New Roman": "OpenSans-Regular.ttf",
  Arial: "OpenSans-Regular.ttf",
  "Open Sans": "OpenSans-Regular.ttf",
};

export function resolveSafeFont(name: string): string | null {
  const filename = SAFE_FONTS[name];
  return filename ? join(FONTS_DIR, filename) : null;
}

export function loadFontMetrics(fontPath: string): FontMetrics {
  const font = openFontSync(fontPath);
  const unitsPerEm = font.unitsPerEm;
  const advanceWidths = new Map<string, number>();

  // Sample the printable ASCII range plus space — mirrors the Python
  // side's cmap-driven per-character advance width table.
  for (let code = 0x20; code <= 0x7e; code++) {
    const char = String.fromCharCode(code);
    const glyph = font.glyphForCodePoint(code);
    if (glyph && glyph.advanceWidth > 0) {
      advanceWidths.set(char, glyph.advanceWidth);
    }
  }

  const spaceAdvance = advanceWidths.get(" ");
  const defaultAdvance = spaceAdvance ?? Math.round(unitsPerEm * 0.5);

  return { advanceWidths, unitsPerEm, defaultAdvance };
}

function charWidthPt(char: string, metrics: FontMetrics, fontSizePt: number): number {
  const raw = metrics.advanceWidths.get(char) ?? metrics.defaultAdvance;
  return (raw / metrics.unitsPerEm) * fontSizePt;
}

export function measureTextWidthPt(text: string, metrics: FontMetrics, fontSizePt: number): number {
  let total = 0;
  for (const char of text) total += charWidthPt(char, metrics, fontSizePt);
  return total;
}

export function wrapLines(
  text: string,
  metrics: FontMetrics,
  fontSizePt: number,
  maxWidthPt: number,
): string[] {
  const words = text.split(/\s+/).filter((w) => w.length > 0);
  const lines: string[] = [];
  let currentWords: string[] = [];
  let currentWidth = 0;
  const spaceW = charWidthPt(" ", metrics, fontSizePt);

  for (const word of words) {
    const wordW = measureTextWidthPt(word, metrics, fontSizePt);
    const candidateWidth = currentWords.length === 0 ? wordW : currentWidth + spaceW + wordW;
    if (currentWords.length > 0 && candidateWidth > maxWidthPt) {
      lines.push(currentWords.join(" "));
      currentWords = [word];
      currentWidth = wordW;
    } else {
      currentWords.push(word);
      currentWidth = candidateWidth;
    }
  }
  if (currentWords.length > 0) lines.push(currentWords.join(" "));
  return lines.length > 0 ? lines : [""];
}

export interface OverflowReport {
  lines: string[];
  lineCount: number;
  totalHeightPt: number;
  boxHeightPt: number;
  overflow: boolean;
}

export function checkOverflow(
  text: string,
  metrics: FontMetrics,
  fontSizePt: number,
  boxWidthPt: number,
  boxHeightPt: number,
  lineHeightFactor = 1.2,
): OverflowReport {
  const lines = wrapLines(text, metrics, fontSizePt, boxWidthPt);
  const totalHeightPt = lines.length * fontSizePt * lineHeightFactor;
  return {
    lines,
    lineCount: lines.length,
    totalHeightPt,
    boxHeightPt,
    overflow: totalHeightPt > boxHeightPt,
  };
}

export function buildOverflowError(
  slide: number,
  primitivePath: string,
  field: string,
  report: OverflowReport,
  fontSizePt: number,
): Record<string, unknown> {
  const overBy = Math.max(0, report.totalHeightPt - report.boxHeightPt);
  return {
    slide,
    primitive: primitivePath,
    field,
    error: "overflow",
    detail: `Text is ~${overBy.toFixed(0)}pt too tall for the box at font size ${fontSizePt}pt (${report.lineCount} lines).`,
    fix: "Shorten the text, reduce bullet/line count, or split into two slides.",
  };
}

// Last-resort fallback if the bundled reference font (SAFE_FONTS) somehow
// fails to load in a given environment. Overflow validation is skipped,
// not faked, when none of these are found either.
const FALLBACK_SYSTEM_FONTS = [
  "C:/Windows/Fonts/arial.ttf",
  "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf",
  "/System/Library/Fonts/Supplemental/Arial.ttf", // macOS
];

/** Pick the font file overflow validation should measure against, for
 * *this* deck's template. v1 decks have exactly one template, so this is a
 * single lookup per validate()/renderDeck() call, not per-slide.
 */
export function resolveFontPath(template: Template): string | null {
  const path = resolveSafeFont(template.fontFamily);
  if (path && existsSync(path)) return path;
  return FALLBACK_SYSTEM_FONTS.find((p) => existsSync(p)) ?? null;
}
